packages/uploadthing/src/client.ts

```diff
@@ -136,58 +166,50 @@ const uploadFile = <
   },
   presigned: PresignedURLs[number],
 ) =>
-  Effect.gen(function* () {
-    const file = opts.files.find((f) => f.name === presigned.fileName);
-
-    if (!file) {
-      console.error("No file found for presigned URL", presigned);
-      return yield* new UploadThingError({
-        code: "NOT_FOUND",
-        message: "No file found for presigned URL",
-        cause: `Expected file with name ${presigned.fileName} but got '${opts.files.join(",")}'`,
-      });
-    }
-
-    opts.onUploadBegin?.({ file: file.name });
-    if ("urls" in presigned) {
-      yield* uploadMultipartWithProgress(file, presigned, opts);
-    } else {
-      yield* uploadPresignedPostWithProgress(file, presigned, opts);
-    }
-
-    const PollingResponse = S.Union(
-      S.Struct({
-        status: S.Literal("done"),
-        callbackData: S.Any as S.Schema<TServerOutput, any>,
-      }),
-      S.Struct({ status: S.Literal("still waiting") }),
-    );
-
-    let serverData: TServerOutput | null = null;
-    if (!opts.skipPolling) {
-      serverData = yield* fetchEffJson(presigned.pollingUrl, PollingResponse, {
+  Array.findFirst(opts.files, (_) => _.name === presigned.fileName).pipe(
+    Effect.tapError(() =>
+      Console.error("No file found for presigned URL", presigned),
+    ),
+    Effect.mapError(
+      () =>
+        new UploadThingError({
+          code: "NOT_FOUND",
+          message: "No file found for presigned URL",
+          cause: `Expected file with name ${presigned.fileName} but got '${opts.files.join(",")}'`,
+        }),
+    ),
+    Effect.tap((_) => opts.onUploadBegin?.({ file: _.name })),
+    Effect.tap((_) =>
+      "urls" in presigned
+        ? uploadMultipartWithProgress(_, presigned, opts)
+        : uploadPresignedPostWithProgress(_, presigned, opts),
+    ),
```

Comment:
I am not with enough context to understand the flow completely, but the pipe form exposes an interesting observation: nearly all our logic is side-effects-- even the main logic (I assume), the upload is something that is done not as part as the main flow. 

Optimally, in my opinion, instead of uploading and then polling in a detached manner like this, this entire thing can be made into a stream, uploads (and downloads) fit streams very naturally. It would allow a shape more similar to
```ts
const uploadFile = (file, opts) => 
  isomorphicUpload(file, opts).pipe(
        Stream.runForEach(progress =>  => {....}), // idk, report it? callback? w/e
        Effect.map(() => ({/* the big obj with the stuff */}))
)
```

The intuition to make the upload an effect is well placed, since it'll finish only once it... finishes! But given the subsequent polling, idk?
You can also concat a polling stream to the upload stream if only polling is possible. Anyways it's just a natural way to express this flow I think